tests/test_weighted_random_sampler.py

```diff
@@ -7,55 +7,104 @@
 pytestmark = [pytest.mark.memcheck]
 
 
-@pytest.mark.parametrize("seed", [2022])
-@pytest.mark.parametrize("size, num_samples", [(500, 1500), (1000, 3000)])
+@pytest.mark.parametrize("seed", [2024])
+@pytest.mark.parametrize("size, num_samples", [(500, 1500), (1000, 3000), (1000, 10000)])
 @pytest.mark.parametrize("precision", [np.float32, np.float64])
-def test_random_sampler(seed, size, num_samples, precision):
+def test_segmented_random_sampler(seed, size, num_samples, precision):
     rng = np.random.default_rng(seed)
 
     # Get a random probability vector.
-    probs = rng.exponential(scale=1 / 250, size=size)
-    prob_total = np.sum(probs)
-    probs /= prob_total
-    # Sort the probabilities to easily visualize
-    probs = np.sort(probs)[::-1]
+    probs = rng.exponential(scale=1 / 250, size=(num_samples, size))
 
-    klass = custom_ops.WeightedRandomSampler_f32
+    klass = custom_ops.SegmentedWeightedRandomSampler_f32
     if precision == np.float64:
-        klass = custom_ops.WeightedRandomSampler_f64
+        klass = custom_ops.SegmentedWeightedRandomSampler_f64
 
     x = np.arange(size)
-    ref_selection = rng.choice(x, size=num_samples, p=probs)
-    sampler = klass(size, seed)
+    ref_selection = np.array([rng.choice(x, size=1, p=batch / np.sum(batch)) for batch in probs]).reshape(-1)
+    sampler = klass(size, num_samples, seed)
 
-    test_selection = sampler.sample(num_samples, probs)
+    test_selection = sampler.sample(probs.tolist())
     assert len(test_selection) == num_samples
 
     ks, pv = ks_2samp(ref_selection, test_selection)
     assert ks < 0.05, (pv, ks)
 
 
-@pytest.mark.parametrize("seed", [2022, 2023])
+@pytest.mark.parametrize("seed", [2024])
+@pytest.mark.parametrize("num_samples", [8000])
 @pytest.mark.parametrize("precision", [np.float32, np.float64])
-def test_random_sampler_simple_distribution(seed, precision):
+def test_segmented_random_sampler_jagged_batches_simple_distributions(seed, num_samples, precision):
+    """Make sure that jagged arrays are handled_correctly"""
+    klass = custom_ops.SegmentedWeightedRandomSampler_f32
+    if precision == np.float64:
+        klass = custom_ops.SegmentedWeightedRandomSampler_f64
+
+    weights = []
```

Comment:
+1 for strengthening test with these cases.

>6. contains a single +inf (okay, since it's a huge weight)

quibble: 6 should also throw an error. (sum(weights) should be positive and finite)

Reply:
> contains a single 0.0 (? not sure what correct behavior here is)

There is already a test for this, should never sample this value. https://github.com/proteneer/timemachine/blob/683d8018e359f0c204abc6c400d6a6cb67fbb790/tests/test_weighted_random_sampler.py#L94

Reply:
Ah I missed it, thanks!

re: quibble, yeah I think it's fine to throw an error if there's a `+inf` - I was thinking about computing the probability of the "reverse" of the hastings proposal factor, which can have an infinite weight (if there's a clash). But we never need to *sample* from the reverse probabilities, so it's okay to throw an error.

Reply:
For these tests do we want asserts in the kernels or just validate these things in the host layers? Asserts will trigger the process to exit so we can't test those in the code as is, but any host layer checks won't occur in practice. Probably say best to add both, but there will be a discontinuity between the host/device code.

Reply:
Changed the total sum of the distributions to be different in https://github.com/proteneer/timemachine/pull/1235/commits/1c574fd6cf00f6868b3bb5d99d2e4ad54c641f5f

Reply:
Added validation for 3-8 of the original comment in https://github.com/proteneer/timemachine/pull/1235/commits/b226056ebcc536eda6f0a37d03bc6ddc96984ce9

Did change some tests that had bad clashes that we wouldn't support in practice.